File: client/src/app/models/rock-explorer-feature.ts
import { AbstractModel } from './abstract-model';
import { User } from './user';
import { RockExplorerPotential } from '../enums/rock-explorer-potential';
import { RockExplorerRockQuality } from '../enums/rock-explorer-rock-quality';
import { RockExplorerRockType } from '../enums/rock-explorer-rock-type';
import { RockExplorerAccessIssue } from '../enums/rock-explorer-access-issue';
import { LineType } from '../enums/line-type';
import { Geometry, LineString } from 'geojson';
import { Tag } from './tag';
import {
  RockExplorerParkingSite,
  RockExplorerPath,
  RockExplorerPathSource,
} from './rock-explorer-misc';

export type RockExplorerFeatureStatus = 'draft' | 'published';

/**
 * Rock Explorer mapped feature (Point or Polygon).
 * Drafts may omit geometry until publish.
 * Pause/resume GPS capture is client session state — not on this model.
 */
export class RockExplorerFeature extends AbstractModel {
  title: string | null;
  description: string | null;
  status: RockExplorerFeatureStatus;
  recordingDeviceId: string | null;
  recordingUpdatedAt: string | null;
  potential: RockExplorerPotential | null;
  rockQuality: RockExplorerRockQuality | null;
  rockType: RockExplorerRockType | null;
  gradeLineType: LineType | null;
  gradeScale: string | null;
  gradeValueMin: number | null;
  gradeValueMax: number | null;
  accessIssues: RockExplorerAccessIssue[] = [];
  geometry: Geometry | null;
  parkingSites: RockExplorerParkingSite[] = [];
  paths: RockExplorerPath[] = [];
  /** Topo targets stored as Tag associations (same shape as gallery tags). */
  topoLinks: Tag[] = [];
  createdBy: User | null;

  public static deserialize(payload: any): RockExplorerFeature {
    const feature = new RockExplorerFeature();
    AbstractModel.deserializeAbstractAttributes(feature, payload);
    feature.title = payload.title ?? null;
    feature.description = payload.description ?? null;
    feature.status = payload.status ?? 'published';
    feature.recordingDeviceId = payload.recordingDeviceId ?? null;
    feature.recordingUpdatedAt = payload.recordingUpdatedAt ?? null;
    feature.potential = payload.potential ?? null;
    feature.rockQuality = payload.rockQuality ?? null;
    feature.rockType = payload.rockType ?? null;
    feature.gradeLineType = payload.gradeLineType ?? null;
    feature.gradeScale = payload.gradeScale ?? null;
    feature.gradeValueMin = payload.gradeValueMin ?? null;
    feature.gradeValueMax = payload.gradeValueMax ?? null;
    feature.accessIssues = payload.accessIssues ?? [];
    feature.geometry = payload.geometry ?? null;
    feature.parkingSites = payload.parkingSites ?? [];
    feature.paths = (payload.paths ?? []).map((path: any) => ({
      id: path.id,
      source: (path.source as RockExplorerPathSource) ?? 'manual',
      title: path.title ?? null,
      description: path.description ?? null,
      geometry: path.geometry as LineString,
    }));
    feature.topoLinks = (payload.topoLinks ?? []).map(Tag.deserialize);
    feature.createdBy = payload.createdBy
      ? User.deserialize(payload.createdBy)
      : null;
    return feature;
  }

  public static serialize(feature: RockExplorerFeature): any {
    const status = feature.status ?? 'published';
    const body: any = {
      title: feature.title,
      description: feature.description,
      status,
      potential: feature.potential,
      rockQuality: feature.rockQuality,
      rockType: feature.rockType,
      gradeLineType: feature.gradeLineType,
      gradeScale: feature.gradeScale,
      gradeValueMin: feature.gradeValueMin,
      gradeValueMax: feature.gradeValueMax,
      accessIssues: feature.accessIssues ?? [],
      geometry: feature.geometry,
      parkingSites: (feature.parkingSites ?? []).filter(
        (site) => site.lat != null && site.lng != null,
      ),
      paths: (feature.paths ?? []).filter(
        (path) => (path.geometry?.coordinates?.length ?? 0) >= 2,
      ),
      topoLinks: (feature.topoLinks ?? []).map(Tag.serialize),
    };
    if (status === 'draft') {
      body.recordingDeviceId = feature.recordingDeviceId;
    } else {
      body.recordingDeviceId = null;
    }
    return body;
  }
}

File: server/src/models/rock_explorer_feature.py
from sqlalchemy import Table
from sqlalchemy.dialects.postgresql import JSON

from extensions import db
from models.base_entity import BaseEntity
from models.enums.line_type_enum import LineTypeEnum
from models.enums.rock_explorer_feature_status_enum import RockExplorerFeatureStatusEnum
from models.enums.rock_explorer_potential_enum import RockExplorerPotentialEnum
from models.enums.rock_explorer_rock_quality_enum import RockExplorerRockQualityEnum
from models.enums.rock_explorer_rock_type_enum import RockExplorerRockTypeEnum
from models.tag import Tag

rock_explorer_feature_tags = Table(
    "rock_explorer_feature_tags",
    db.metadata,
    db.Column("tag_id", db.ForeignKey("tags.id")),
    db.Column("rock_explorer_feature_id", db.ForeignKey("rock_explorer_features.id")),
)


class RockExplorerFeature(BaseEntity):
    """
    A mapped exploration feature (Point or Polygon GeoJSON geometry).
    """

    __tablename__ = "rock_explorer_features"

    geometry = db.Column(JSON, nullable=True)
    parking_sites = db.Column(JSON, nullable=False, default=lambda: [])
    paths = db.Column(JSON, nullable=False, default=lambda: [])

    title = db.Column(db.String(120), nullable=True)
    description = db.Column(db.Text, nullable=True)
    potential = db.Column(db.Enum(RockExplorerPotentialEnum), nullable=True)
    rock_quality = db.Column(db.Enum(RockExplorerRockQualityEnum), nullable=True)
    rock_type = db.Column(db.Enum(RockExplorerRockTypeEnum), nullable=True)
    grade_line_type = db.Column(db.Enum(LineTypeEnum), nullable=True)
    grade_scale = db.Column(db.String(32), nullable=True)
    grade_value_min = db.Column(db.Integer, nullable=True)
    grade_value_max = db.Column(db.Integer, nullable=True)
    access_issues = db.Column(JSON, nullable=False, default=lambda: [])

    status = db.Column(
        db.Enum(
            RockExplorerFeatureStatusEnum,
            values_callable=lambda enum_cls: [member.value for member in enum_cls],
        ),
        nullable=False,
        default=RockExplorerFeatureStatusEnum.PUBLISHED,
    )
    recording_device_id = db.Column(db.String(128), nullable=True)
    recording_updated_at = db.Column(db.DateTime(), nullable=True)

    # Same Tag association pattern as GalleryImage.tags (topo links).
    topo_links = db.relationship(Tag, secondary=rock_explorer_feature_tags)
